app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/ExecutionAndStagePicker.tsx

```diff
@@ -0,0 +1,144 @@
+import * as React from 'react';
+import { keyBy, isEqual } from 'lodash';
+import { Option } from 'react-select';
+
+import { IExecution, IPipeline, IStage } from 'core/domain';
+import { FormField, IStageForSpelPreview, ReactSelectInput, useData } from 'core/presentation';
+import { ReactInjector } from 'core/reactShims';
+import { timestamp, relativeTime } from 'core/utils';
+import { ExecutionStatus } from 'core/pipeline/status/ExecutionStatus';
+
+export interface IExecutionAndStagePickerProps {
+  pipeline: IPipeline;
+  pipelineStage: IStage;
+  onChange(chosenExecutionAndStage: IStageForSpelPreview): void;
+}
+
+export function ExecutionAndStagePicker(props: IExecutionAndStagePickerProps) {
+  const { pipeline, pipelineStage, onChange } = props;
+  const { executionService } = ReactInjector;
+  const [execution, setExecution] = React.useState<IExecution>(null);
+  const [stageId, setStageId] = React.useState<string>(null);
+  const [showStageSelector, setShowStageSelector] = React.useState(false);
+
+  const fetchExecutions = useData(
+    () => executionService.getExecutionsForConfigIds([pipeline.id], { limit: 100 }),
+    [],
+    [],
+  );
+  const executions = fetchExecutions.result;
+  const stages = execution && execution.stages;
+
+  // When executions load, select the first one.
+  React.useEffect(() => executions && setExecution(executions[0]), [executions]);
+
+  // When an execution is chosen or the current pipeline or pipeline stage changes,
+  // find the matching stage from the execution
+  React.useEffect(() => {
+    const exactStage = findExactStageFromExecution(pipeline, execution, pipelineStage);
+    setShowStageSelector(!!exactStage);
+    setStageId(exactStage || findCloseStageFromExecution(pipeline, execution, pipelineStage));
+  }, [pipeline, pipelineStage, execution]);
+
+  // When the chosen execution or stage changes, notify the parent
+  React.useEffect(() => {
+    onChange({ executionLabel: executionLabel(execution), executionId: execution && execution.id, stageId });
+  }, [execution, stageId]);
+
+  const stageOptions: Option[] = stages ? stages.map(x => ({ label: x.name, value: x.id })) : [];
+  const executionOptions: Array<Option<IExecution>> = executions ? executions.map(value => ({ value })) : [];
```

Comment:
what if there are no prior executions? is there some helptext/indication that preview won't work

Reply:
+1 to some help text when there are no prior executions if possible!